src/integrationTest/java/com/hivemq/helmcharts/util/OperatorHelmChartContainer.java

```diff
@@ -1,85 +1,164 @@
-        final String s = new ImageFromDockerfile().withDockerfile(
-                        dockerfile.toPath())
+    private static @NotNull DockerImageName getAdHocImageName(final @NotNull String k3sVersion,
+                                                             final @NotNull String dockerfileName) {
+        var dockerfile = new File(MountableFile.forClasspathResource(dockerfileName).getFilesystemPath());
+
+        final String s = new ImageFromDockerfile().withDockerfile(dockerfile.toPath())
                 .withBuildArg("K3S_VERSION", k3sVersion)
                 .get();
-        return DockerImageName.parse(s).asCompatibleSubstituteFor("rancher/k3s");
-    }
 
-    public static @NotNull Builder builder() {
-        return new Builder();
+        return DockerImageName.parse(s).asCompatibleSubstituteFor("rancher/k3s");
     }
 
-    public int getMappedPort() {
-        return super.getMappedPort(mqttPort);
+    /**
+     * Uses custom images instead of docker hub images, additional images can be added
+     */
+    public @NotNull OperatorHelmChartContainer withCustomImages(@Nullable String... fileNames) {
+        withCustomImages = true;
+        super.withFileSystemBind("./build/containers", "/build", BindMode.READ_ONLY);
+        imagesNames.addAll(Arrays.asList("hivemq-init-dns-image.tar", "hivemq-k8s-image.tar", "hivemq-operator.tar"));
+        imagesNames.addAll(Arrays.asList(fileNames));
+        return this;
     }
 
-    public void waitForMqttService() {
-        super.withExposedPorts(mqttPort).waitingFor(Wait.forListeningPort());
+    protected @NotNull List<String> getImagesNames() {
+        return imagesNames;
     }
 
-    public static class Builder {
-
-        private @Nullable File dockerfile;
-        private @Nullable File helmChartMountPath;
-        private @Nullable String k3sVersion;
-        private @Nullable String containerPath;
+    private static class DeploymentStatusStartupCheckStrategy extends StartupCheckStrategy {
+        private final @NotNull OperatorHelmChartContainer container;
 
-        public @NotNull Builder dockerfile(final @NotNull File dockerfile) {
-            this.dockerfile = dockerfile;
-            return this;
+        public DeploymentStatusStartupCheckStrategy(@NotNull OperatorHelmChartContainer container) {
+            this.container = container;
+            this.withTimeout(Duration.ofSeconds(240));
         }
 
-        public @NotNull Builder helmChartMountPath(final @NotNull File mountPath) {
-            this.helmChartMountPath = mountPath;
-            return this;
+        @Override
+        public @NotNull StartupStatus checkStartupState(@NotNull DockerClient dockerClient, @NotNull String containerId) {
+            var s = container.getLogs(STDERR);
+            while (!s.matches("(?s).*Node controller sync successful.*")) {
+                s = container.getLogs(STDERR);
+            }
```

Comment:
Replace this spinning log with https://mvnrepository.com/artifact/org.awaitility/awaitility or use https://github.com/hivemq/hivemq-security-extension/blob/6415475e151c2a7818768a8b75b3716838261a47/src/integrationTest/java/com/hivemq/extensions/security/util/LogWaiterUtil.java